sunpy/net/dataretriever/sources/goes.py

```diff
@@ -52,33 +89,80 @@ def _get_url_for_timerange(self, timerange, **kwargs):
+            return result
+
+        def download_irradiance(time_range):
+            now = datetime.datetime.now()
+            real_time = ((now - time_range.end).total_seconds() <= 300)
+            result = list()
+            total_days = (time_range.end - time_range.start).days + 1
+            all_dates = time_range.split(total_days)
+
+            g = lambda url, char, date, cad: url.format(char=char, date= date, cad=cad)
+            url_pattern = 'ftp://ftp.swpc.noaa.gov/pub/lists/xray/{date:%Y%m%d}_G{char}_xr_{cad}m.txt'
+            for day in all_dates:
+                result.extend([g(url_pattern, 's', day.end, cad) for cad in (1, 5)])
+                result.extend([g(url_pattern, 'p', day.end, cad) for cad in (1, 5)])
+            if real_time:
+                result.extend(['ftp://ftp.swpc.noaa.gov/pub/lists/xray/G{char}_xr_1m.txt'.format(char=ch) for ch in ('p', 's')])
+                result.extend(['ftp://ftp.swpc.noaa.gov/pub/lists/xray/G{char}_xr_5m.txt'.format(char=ch) for ch in ('p', 's')])
+            return result
+
+        def download_intensity(time_range):
+            prefix = 'http://satdat.ngdc.noaa.gov/sxi/archive/fits/goes15/{date:%Y/%m/%d/}'
+            suffix = ['SXI_%Y%m%d_%H%M%S%j_{pat}_15.FTS'.format(pat=pat) for pat in ('AB', 'BA')]
+            result = list()
+            total_days = (time_range.end - time_range.start).days + 1
+            all_dates = time_range.split(total_days)
+            for day in all_dates:
+                html = urllib2.urlopen(prefix.format(date= day.end))
+                soup = BeautifulSoup(html)
+                for link in soup.findAll("a"):
+                    url = str(link.get('href'))
+                    for suf in suffix:
+                        crawler = Scraper(suf)
+                        if crawler._URL_followsPattern(url):
+                            extract = crawler._extractDateURL(url)
+                            valid = datetime.datetime(day.end.year, day.end.month, day.end.day,
+                                                      extract.hour, extract.minute, extract.second)
+                            if time_range.start <= valid <= time_range.end:
+                                result.append(prefix.format(date= day.end) + url)
+            return result
+
+        start_dates = {'INTENSITY': datetime.datetime(2010, 6, 4), 'IRRADIANCE':datetime.datetime(2012, 9, 6),
+                       'PARTICLE_FLUX': datetime.datetime(2015, 5, 30)}
+        START_DATE = start_dates[physobs]
+        if timerange.start < START_DATE:
+            raise ValueError('Earliest date for which data is available is {:%Y-%m-%d}'.format(START_DATE))
+
+        if physobs == 'INTENSITY':
+            result = download_intensity(timerange)
+        elif physobs == 'PARTICLE_FLUX':
+            result = download_particle_flux(timerange)
+        else:
+            result = download_irradiance(timerange)
```

Comment:
That was the first thing I tried. But it would depend if python does lazy-evaluation or not. If it does then we're clear, if it **doesn't** then it would needlessly compute all functions in the dictionary which would mean unnecessarily download all fits files and hence would slow down the program.